benchbuild/utils/requirements.py

```diff
@@ -0,0 +1,361 @@
+import logging
+import typing as tp
+from enum import Enum
+import copy
+import abc
+import attr
+
+from benchbuild.settings import CFG
+
+LOG = logging.getLogger(__name__)
+
+RequirementSubType = tp.TypeVar("RequirementSubType", bound='Requirement')
+
+
+@attr.s
+class Requirement:
+    """
+    Base class for requirements.
+    """
+    @abc.abstractmethod
+    def to_option(self) -> str:
+        """
+        Converts Requirement to a script options.
+        """
+
+    @abc.abstractmethod
+    def to_cli_option(self) -> str:
+        """
+        Converts Requirement to a command line options.
+        """
+
+    @classmethod
+    @abc.abstractmethod
+    def merge_requirements(
+            cls: tp.Type[RequirementSubType], lhs_option: RequirementSubType,
+            rhs_option: RequirementSubType) -> RequirementSubType:
+        """
+        Merge the requirements of the same type together.
+        """
+        return type(lhs_option).merge_requirements(lhs_option, rhs_option)
+
+
+################################################################################
+# Slurm Requirements                                                           #
+################################################################################
+
+
+class SlurmRequirement(Requirement):
+    """
+    Base class for slurm requirements.
+    """
+    def to_option(self) -> str:
+        """
+        Converts Requirement to a script options.
+        """
+        return self.to_slurm_opt()
+
+    def to_cli_option(self) -> str:
+        """
+        Converts Requirement to a command line options.
+        """
+        return self.to_slurm_cli_opt()
+
+    def to_slurm_opt(self) -> str:
+        """
+        Converst slurm option into a script usable option string, i.e., bash
+        #SBATCH option line.
+        """
+        return f"#SBATCH {self.to_slurm_cli_opt()}"
+
+    @abc.abstractmethod
+    def to_slurm_cli_opt(self) -> str:
+        """
+        Converst slurm option to command line string.
+        """
+
+
+@attr.s
+class SlurmCoresPerSocket(SlurmRequirement):
+    """
+    Restrict node selection to nodes with at least the specified number of
+    cores per socket. See additional information under -B option above when
```

Comment:
above what?

Reply:
these are from the slurm docs, I'll updated the wording